src/dart_document_symbol_provider.ts

```diff
@@ -0,0 +1,64 @@
+"use strict";
+
+import { TextDocument, DocumentSymbolProvider, SymbolInformation, CancellationToken, SymbolKind, Location, Uri, Range, Position } from "vscode";
+import { Analyzer, getSymbolKindForElementKind } from "./analyzer";
+import { toRange } from "./utils";
+import * as as from "./analysis_server_types";
+
+export class DartDocumentSymbolProvider implements DocumentSymbolProvider {
+	private analyzer: Analyzer;
+
+	constructor(analyzer: Analyzer) {
+		this.analyzer = analyzer;
+	}
+
+	provideDocumentSymbols(document: TextDocument, token: CancellationToken): Thenable<SymbolInformation[]> {
+		let file = document.fileName;
+
+		this.analyzer.analysisSetSubscriptions({
+			subscriptions: { 'OUTLINE': [file] }
+		});
+
+		// TODO: Improve the presentation to make the file structure clearer?
+
+		return new Promise<SymbolInformation[]>((resolve, reject) => {
+			let disposable = this.analyzer.registerForAnalysisOutline(n => {
+				if (n.file != file)
+					return;
+
+				this.analyzer.analysisSetSubscriptions({
+					subscriptions: { 'OUTLINE': [] }
+				});
+				disposable.dispose();
+
+				let symbols: SymbolInformation[] = [];
+				for (let element of n.outline.children)
+					this.transcribeOutline(symbols, null, element);
+				resolve(symbols);
+			});
+		});
+	}
+
+	private transcribeOutline(symbols: SymbolInformation[], parent, outline) {
+		let element = outline.element;
+
+		let name = element.name;
+		if (element.kind == "CONSTRUCTOR" || element.kind == "METHOD" || element.kind == "FUNCTION")
+			name = `${name}()`;
```

Comment:
In [the docs](https://htmlpreview.github.io/?https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/doc/api.html#type_Element) there's a `parameters` property which seems it eliminate the need for this if but also provide the parameter list?

> **parameters ( optional String )**
> The parameter list for the element. If the element is not a method or function this field will not be defined. If the element doesn't have parameters (e.g. getter), this field will not be defined. If the element has zero parameters, this field will have a value of "()".